wasm/derive/src/lib.rs

```diff
@@ -1,89 +1,229 @@
+    };
     let fn_name = &sig.ident;
 
     block.stmts.insert(
         0,
         parse_quote!(
-            use iroha_wasm::Execute as _;
+            use ::iroha_wasm::Execute as _;
         ),
     );
 
     quote! {
-        // NOTE: The size of the `len` parameter is defined by the target architecture
-        // which is `wasm32-unknown-unknown` and therefore not dependent by the architecture
-        // smart contract is compiled on or the architecture smart contract is run on
         /// Smart contract entry point
-        ///
-        /// # Safety
-        ///
-        /// Given pointer and length must comprise a valid memory slice
         #[no_mangle]
-        pub unsafe extern "C" fn _iroha_wasm_main(ptr: *const u8, len: usize) {
-            #fn_name(iroha_wasm::_decode_from_raw::<<Account as Identifiable>::Id>(ptr, len))
+        pub unsafe extern "C" fn _iroha_trigger_main(
+        ) {
+            #fn_name(#args)
         }
 
-        #[allow(clippy::needless_pass_by_value)]
         #(#attrs)*
         #vis #sig
         #block
     }
     .into()
 }
 
-fn verify_function_signature(sig: &Signature) {
-    if ReturnType::Default != sig.output {
-        abort!(sig.output, "Exported function must not have a return type");
-    }
-
-    if sig.inputs.len() != 1 {
-        abort!(
-            sig.inputs,
-            "Exported function must have exactly 1 input argument of type `Account::Id`"
-        );
-    }
-
-    if let Some(syn::FnArg::Typed(pat)) = sig.inputs.iter().next() {
-        if !type_is_account_id(&pat.ty) {
-            abort!(
-                pat.ty,
-                "Argument to the exported function must be of the `Account::Id` type"
-            );
-        }
-    }
-}
-
-fn type_is_account_id(account_id_ty: &Type) -> bool {
-    if *account_id_ty == parse_quote!(<Account as Identifiable>::Id) {
-        return true;
-    }
-
-    if let Type::Path(path) = account_id_ty {
-        let Path { segments, .. } = &path.path;
-
-        if let Some(type_name) = segments.last().map(|ty| &ty.ident) {
-            return *type_name == "AccountId";
-        }
-    }
+fn construct_args(
+    types: &Punctuated<ParamType, syn::token::Comma>,
+) -> Punctuated<syn::Expr, syn::token::Comma> {
+    types
+        .iter()
+        .map(|param_type| -> syn::Expr {
+            match param_type {
+                ParamType::Authority => {
+                    parse_quote! {
+                        ::iroha_wasm::query_authority()
+                    }
+                }
+                ParamType::TriggeringEvent => {
+                    parse_quote! {{
+                        use ::iroha_wasm::debug::DebugExpectExt as _;
 
-    false
+                        let top_event = ::iroha_wasm::query_triggering_event();
+                        ::core::convert::TryInto::try_into(top_event)
+                            .dbg_expect("Failed to convert top-level event to the concrete one")
```

Comment:
would it be possible to do the conversion before calling the smart contract, not inside of it. I wouldn't mind too much, had it not made you mess up the dbg macro. Also, it's better to move out of smart contract as much code as possible

Reply:
what you're saying is that executing a smart contract is valid only for a certain type of event. If you registered this smart contract for some other event the execution would fail. 

My questions are:
1. Why is it not ok for a smart contract to be able to be triggered on multiple events?
2. Is this covered in the validation step? I wouldn't like to see that we've successfully registered a trigger but when the actual event occurs the execution of smart contract fails (I'm not sure how trigger validation works).

Reply:
> Is this covered in the validation step? I wouldn't like to see that we've successfully registered a trigger but when the actual event occurs the execution of smart contract fails (I'm not sure how trigger validation works).

No yet covered. But yes, I agree, I'd like to get the error about incompatible event type at `Register` step. This is a subject for the next PR, as we discussed with @appetrosyan.

> Why is it not ok for a smart contract to be able to be triggered on multiple events?

There are two notes:
1. Currently we provide `Filter` at trigger registration step. `Filter` can work only with one concrete type of events. That's why our trigger can't receive different event types for now. Maybe this will be changed once we have better filters. That's probably a question for @appetrosyan .
2. You still can use top-level event, because there is always `impl TryInto<T> for T`.

Reply:
Alternative would be to support only `Event` as `entrypoint` argument and leave it to the writer of smart contract to make the conversion. Not sure if I would prefer it, but it kinda makes sense because you are essentially faking strong typing in the `entrypoint` by implicitly doing the conversion

Reply:
Well, I think, all strong-typing with WASM here actually would be faking. You can always send something different instead of i.e. `AccountId` and you won't notice that until runtime.
Here is the same thing. How can this typing be stronger (especially with *pull*-architecture).
In terms of implicit conversion. I think, your approach will allow only one way -- use top-level event. And with my approach user can use any compatible event, including top-level

Reply:
Let's hear more voices here. Democracy!

Reply:
Guys, the typing arguments are all well and good, but we should agree on a few things:

1. We need this soon. ASAP. 
2. If we enforce top-level events first and then keeping the API the same, change the underlying behaviour, we can have a smooth transition. 
3. Allowing anything but the top-level events is UB. We can define that behaviour later.

Reply:
I don't fully understand, what do you mean, but okay. If we need this ASAP, when we need just one more approval here